alpa/pipeline_parallel/cross_mesh_resharding.py

```diff
@@ -1140,6 +1228,231 @@ def _generate_send_recv_resharding_strategy_by_loads(
+            for receiver in dst_tile.replica_device_strs:
+                for j, src_tileslice in enumerate(src_tileslices):
+                    senders = src_tileslice.replica_device_strs
+                    data_size = src_tileslice.tile_size
+                    works.append(
+                        SingleReshardingLoadBalancingWork(
+                            senders, [receiver], data_size))
+
+        task = ReshardingLoadBalancingTaskSolver(src_mesh_devices,
+                                                 dst_mesh_devices,
+                                                 device_host_map, works,
+                                                 nic_constraints)
+
+        sol_assigned_sender, sol_order = task.solve()
+
+        per_spec_plans = []
+        rank_to_idx = {}
+        cnt = 0
+        for i, (dst_tile, src_tileslices,
+                _) in enumerate(spec.dst_tile_to_src_tiles_map):
+            per_spec_plan = np.empty(
+                (len(dst_tile.replica_device_strs), len(src_tileslices)),
+                dtype=object)
+            for k, receiver in enumerate(dst_tile.replica_device_strs):
+                for j, src_tileslice in enumerate(src_tileslices):
+                    sender = sol_assigned_sender[cnt]
+                    per_spec_plan[k][j] = sender
+                    rank_to_idx[cnt] = (i, k, j)
+                    cnt += 1
+            per_spec_plans.append(per_spec_plan)
+
+        order = [rank_to_idx[i] for i in sol_order]
+        is_local_allgather = spec.final_dst_spec != spec.dst_sharding_spec
+        strategy = SendRecvReshardingStrategyWithOrder(per_spec_plans,
+                                                       is_local_allgather,
+                                                       order)
+        return strategy
+
+    @staticmethod
+    def _generate_broadcast_resharding_strategy_by_no_load(spec):
+        """
+            Generate the broadcast-based resharding strategy by balancing
+            loads. For each tile, I not only allow one source to provide
+            the tile.
+        """
+        # pylint: disable=unused-argument
+        per_spec_plans = []
+        for _, src_tileslices, _ in spec.dst_tile_to_src_tiles_map:
+            per_spec_plan = np.empty((len(src_tileslices),), dtype=object)
+
+            for src_tileslice_idx, src_tileslice in enumerate(src_tileslices):
+                per_spec_plan[
+                    src_tileslice_idx] = src_tileslice.replica_device_strs[0]
+            per_spec_plans.append(per_spec_plan)
+        strategy = ReshardingStrategy(per_spec_plans, None)
+        return strategy
+
+    @staticmethod
+    def _generate_broadcast_resharding_strategy_by_loadbalance(
+            spec, src_mesh, dst_mesh):
+        """
+            Generate the broadcast-based resharding strategy by balancing
+            loads and along time.
+        """
+
+        def device_id_to_str(host_ip, device_id, device_type="gpu"):
```

Comment:
this pre-processing(L1392-1419) seems the same as what in `_generate_send_recv_resharding_strategy_by_loadbalance` (L1296-1322), can we use a function to let them share the code?

Reply:
OK